lib/FrontendTool/FrontendTool.cpp

```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
   return SF;
 }
 
+/// Dumps the AST of all available primary source files. If corresponding output
+/// files were specified, use them; otherwise, dump the AST to stderr.
+static void dumpAST(CompilerInstance &Instance) {
+  for(SourceFile *sourceFile: Instance.getPrimarySourceFiles()) {
```

Comment:
Nitpick: We put a space between the `for` and the open paren.

Reply:
Can you also add a FIXME comment that says this doesn't do the right thing for WMO? We don't actually have to fix that right now, since no one is asking for it, but it's good to call out deficiencies like that.

Reply:
What would the right thing be for WMO? You said there aren't any primary files, so could we check for that case and maybe dump all... "non-primary" files?

Reply:
I think `forEachInputProducingAMainOutputFile` in FrontendInputsAndOutputs is probably closest to the right thing, since that's where you get output filenames from. It's *still* not quite going to be right for single-threaded WMO because you only get one output file in that case, which'll be repeatedly overwritten the way the code is written now, but that's even more marginal. Both Xcode and SwiftPM always build with multi-threaded WMO these days.

Reply:
Actually, that doesn't give you SourceFiles. I guess when `getPrimarySourceFiles` is empty, you could iterate over all files from `getMainModule`. This doesn't handle single-threaded WMO either for the reason listed above, which you can special-case or not.

I do still think it's okay to ignore this whole mess for now because you don't need it.

Reply:
Oh, another option that might be slightly more palatable than "ignore this whole mess" is "have the driver ignore WMO options when using -dump-ast, and warn about it". We have similar things where -wmo overrides -incremental.

Reply:
Hmmm I like this idea. But maybe for a future pull request.